### ISSUE TITLE
odbc bind error

### ISSUE THREAD

**Issue by tsxiaofang** [SOCI/soci#763]
1、bind error
when coding as after code:

	values bvs;
	bvs.set(p.first_name);
	bvs.set(p.last_name);
	bvs.set(p.age);

        statement st = (sql.prepare << "insert into Person(first_name, last_name, age) values(:first_name, :last_name, :age)");

	st.bind(bvs);
	st.execute(false);

error code is: void odbc_standard_use_type_backend::bind_by_pos

modify 
       position_ = position++;
to
       position_ = ++position;

2、recommended to add **get_parameter_name_count** interface to **statement_backend**
     because std::string get_parameter_name(int index)  don't know the number of names_

**Comment by mloskot** [comment]
@tsxiaofang Thanks for the report.

Seems like a perfect issue for a simple pull request. Please, consider submitting a PR, if a basic test is included, it will also be also checked.

/cc @vadz

**Comment by vadz** [comment]
I don't use `values`-based API, so I could be missing something, but the proposed change to `bind_by_pos()` doesn't seem right to me, both because the current code looks fine and because this would make ODBC backend inconsistent with all the other ones.

What happens exactly when you run the code above? I.e. what does "bind error" mean?

**Comment by tsxiaofang** [comment]
becuse ODBC  bind_by_pos start number is 0, bind_by_name start number is 1

**Comment by tsxiaofang** [comment]
another issue is bind_by_name like this:

std::string first_name="Bob";
sql<< "insert into Person(first_name, last_name, age) values(:first_name, :first_name, :age)",use(first_name, "first_name");

**Comment by vadz** [comment]
Yes, sorry, I take my words back: the current code indeed seems wrong as `SQLBindParameter()` parameter number argument counts from 1, not 0, while SOCI parameter positions start from 0 (of course).

So the minimal fix proposed above should work but I really don't have any time to test it -- and we definitely need a unit test verifying that this works.

FWIW personally I'd probably prefer to keep SOCI `position_` 0-based and add an explicit `+1` to it when calling `SQLBindParameter()`.

**Comment by mloskot** [comment]
If someone submits PR with fix and test case for the problem soonish (a week), I'll merge it into release/4.0 branch and release with 4.0.0 RC2

**Comment by tsxiaofang** [comment]
I am writing a framework, like mybatis, which used soci. I need to bind the input parameter list, so I need to use bind values. In addition, I need to know the number of names_ from the statement_backend interface, so I need to add an interface to get the number of names_.
